onnxruntime/core/mlas/lib/amd64/QLinearBinaryOpKernelAvx2.asm

```diff
@@ -0,0 +1,436 @@
+ELSE
+        movzx   TargetReg64,BYTE PTR Source
+ENDIF
+        movq    TargetXmm,TargetReg64
+        vpbroadcastd TargetYmm,TargetXmm
+        ENDM
+
+
+;
+; Macro Description:
+;
+;   This macro generates code to unpack 8 x (s/u)int8 to 8 x int32,
+;   according to signed/unsigned.
+;
+; Arguments:
+;
+;   Source - Supplies address of 8 x 8bits integers.
+;
+;   Target - target ymm register.
+;
+;   DataType - S8 or U8
+;
+
+UnpackBytesDWords MACRO Target, Source, DataType
+IFIDN <DataType>, <S8> 
+        vpmovsxbd Target,QWORD PTR Source
+ELSE
+        vpmovzxbd Target,QWORD PTR Source
+ENDIF
+        ENDM
+
+;
+; Macro Description:
+;
+;   This macro generates code to set Target 64bits register with the
+;   max value of signed/unsigned int8 specified by DataType.
+;
+; Arguments:
+;
+;   Target - target 64bits register.
+;
+;   DataType - S8 or U8.
+;
+
+SetMax8BitsValue MACRO Target, DataType
+IFIDN <DataType>, <S8>
+        mov     Target,QWORD PTR 127
+ELSE
+        mov     Target,QWORD PTR 255
+ENDIF
+        ENDM
+
+;
+; Macro Description:
+;
+;   This macro generates code to set Target 64bits register with the
+;   min value of signed/unsigned int8 specified by DataType.
+;
+; Arguments:
+;
+;   Target - target 64bits register.
+;
+;   DataType - S8 or U8.
+;
+
+SetMin8BitsValue MACRO Target, DataType
+IFIDN <DataType>, <S8>
+        mov     Target,QWORD PTR -128
+ELSE
+        mov     Target,QWORD PTR 0
+ENDIF
+        ENDM
+
+;
+; Macro Description:
+;
+;   This macro generates code for function QLinearOpName()
+;   on the specified signed/unsigned int8 DataType.
+;
+; Arguments:
+;
+;   DataType - S8 or U8.
+;
+;   OpName - Name of the QLinearOp, like Add, Mul, etc.
+;
+;   OpInstruction - the assembly code prefix which op() two ymm vector of floats,
+;                   like vaddps, vmulps, etc
+;
+
+QLinearBinaryOpAvx2 MACRO DataType, OpName, OpInstruction
+
+;
+; Routine Description:
+;
+;    This routine implements the kernels for the Quantize Linear OpName
+;    for element type DataType, vector on vector.
+;
+; Arguments:
+;
+;    InputA (rcx) - Supplies the address of InputA.
+;
+;    ScaleA (xmm1) - Supplies A's Scale value in float.
+;
+;    ZeroPointA (r8) - Supplies A's zero point value.
+;
+;    InputB (r9) - Supplies the address of InputB.
+;
+;    ScaleB - Supplies B's Scale value in float.
+;
+;    ZeroPointB - Supplies B's zero point value.
+;
+;    ScaleC - Supplies C's Scale value in float.
+;
+;    ZeroPointC - Supplies C's zero point value.
+;
+;    OutputC - Supplies the address of OutputC.
+;
+;    LengthA - Supplies the number of elements of InputA.
+;
+;    LengthB - Supplies the number of elements of InputB.
+;              LengthB == LengthA or LengthA == 1 or LengthB == 1.
+;
+; Return Value:
+;
+;    None.
+;
+
+        NESTED_ENTRY MlasQLinear&OpName&&DataType&KernelAvx2, _TEXT
+
+        alloc_stack (QLinearBinaryOpFrame.ReturnAddress)
+
+        save_xmm128_avx xmm6,QLinearBinaryOpFrame.SavedXmm6
```

Comment:
I've been slowly removing save_xmm128_avx and just using save_xmm128. The save_xmm128_avx was my own wrapper and isn't adding any value. Can you switch?

Reply:
Sure!